chainstate/types/src/block_status.rs

```diff
@@ -13,48 +13,335 @@
+    // Note: it's better to keep this function private if possible.
+    fn set_last_valid_stage(&mut self, stage: BlockValidationStage) {
+        self.set_field(BlockStatusBitArea::ValidationStage, stage as u64)
+    }
+
+    /// Return true, if the block hasn't been marked as invalid *yet*.
+    pub fn is_ok(&self) -> bool {
+        !(self.validation_failed() || self.has_invalid_parent())
+    }
+
+    /// Return true, if the block is actually valid.
+    pub fn is_fully_valid(&self) -> bool {
+        self.last_valid_stage() == BlockValidationStage::FullyChecked && self.is_ok()
+    }
+
+    pub fn set_validation_failed(&mut self) {
+        self.set_field(BlockStatusBitArea::ValidationFailedBit, 1)
+    }
+
+    pub fn validation_failed(&self) -> bool {
+        self.get_field(BlockStatusBitArea::ValidationFailedBit) != 0
+    }
+
+    pub fn set_has_invalid_parent(&mut self) {
+        self.set_field(BlockStatusBitArea::InvalidParentBit, 1)
+    }
+
+    pub fn has_invalid_parent(&self) -> bool {
+        self.get_field(BlockStatusBitArea::InvalidParentBit) != 0
+    }
+
+    #[cfg(test)]
+    fn reserved_bits(&self) -> u64 {
+        self.get_field(BlockStatusBitArea::ReservedArea)
+    }
+
+    fn bit_range_of(bits: BlockStatusBitArea) -> Range<usize> {
+        bits as usize..bits.next().expect("Can't determine field's end") as usize
+    }
+
+    fn get_field(&self, bits: BlockStatusBitArea) -> u64 {
+        let range = Self::bit_range_of(bits);
+        let lsb_mask = (1 << range.len()) - 1;
+
+        (self.0 >> range.start) & lsb_mask
+    }
+
+    fn set_field(&mut self, bits: BlockStatusBitArea, val: u64) {
+        let range = Self::bit_range_of(bits);
+        let lsb_mask = (1 << range.len()) - 1;
+        let mask = lsb_mask << range.start;
+
+        assert!(val <= lsb_mask);
+        self.0 = (self.0 & !mask) | (val << range.start);
+    }
+}
+
+impl std::fmt::Display for BlockStatus {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "BlockStatus(last_valid_stage: {}, validation_failed: {}, has_invalid_parent: {})",
+            self.last_valid_stage(),
+            self.validation_failed(),
+            self.has_invalid_parent()
+        )
+    }
+}
+
+// Each value here represents the bit index where the corresponding field starts; the end of the
+// field will be the start of the next one. E.g. below "validation stage" will occupy bits [0, 8),
+// "validation failed bit" will occupy bits [8, 9) (i.e. the single 8th bit) etc.
+#[derive(Sequence, Clone, Copy)]
+enum BlockStatusBitArea {
+    ValidationStage = 0,
+    ValidationFailedBit = 8,
+    InvalidParentBit,
+    ReservedArea,
+    End = 64,
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::panic::catch_unwind;
+
+    #[test]
+    fn test_bit_range_of() {
+        // FIXME: each of the asserts below is below 100 characters in length, so they could be
```

Comment:
You can put `#[rustfmt::skip]` on the function, and it'll be skipped.

Reply:
> You can put #[rustfmt::skip] on the function, and it'll be skipped.

[discussed] I don't really want to fix it for this particular code, but rather to consider improving the formatter settings so that all other code becomes more compact as well.

> If we can experiment and see how it plays out on the codebase without stucking in formatting debates I'd give it a try.

I've created a throw-away PR, which roughly illustrates how the code might change - https://github.com/mintlayer/mintlayer-core/pull/1038